File: Counting Valleys.cpp
#include <iostream>
using namespace std;

int main(){
    int n;
    int start = 0;
    int count = 0;
    cin >> n;
    char arr[n];
    int temp[n];
    for(int i=0;i<n;i++){
        cin >> arr[i];
    }
    
    for(int i=0;i<n;i++){
        if(arr[i] == 'U'){
            start += 1;
            temp[i] = start;
        }
        else{
            start -= 1;
            temp[i] = start;
        }
        if(start == 0 && temp[i-1] == -1){
            count++;
        }
    }
    
    cout << count << endl;
    return 0;
}

File: Python solutions/Counting valleys.py
n = int(input())
arr = input()
start = 0
count = 0
temp = []

for i in range(0,len(arr)):
    if arr[i] == 'U':
        start += 1
        temp.append(start)
    else:
        start -= 1
        temp.append(start)
    if start == 0 and temp[i-1] == -1:
        count += 1
        
print(count)
